fix(components): Changing modal close button to use <Button> component

kingstonfung opened the pull request:
<!--
  Atlantis uses Conventional Commits to track versions.
  Pull request titles should follow the following format.

  <TYPE>(<optional SCOPE>): <conditionally BREAKING CHANGE:> <description>

  eg.
    fix(SCOPE): stop graphite breaking when too much pressure applied — Patch Release
    feat(SCOPE): add 'graphiteWidth' option — (Minor) Feature Release
    feat(SCOPE): BREAKING CHANGE: remove graphiteWidth option — (Major) Breaking Release

  TYPE should consist of:
    - fix: a commit of the type fix patches a bug in your codebase
    - feat: a commit of the type feat introduces a new feature to the codebase
    - docs: documentation only changes
    - build: improvements to the build system
    - refactor: a change that neither fixes a bug nor introduces a feature
    - chore: other changes that don't modify src or test files

  SCOPE should be one of:
    - components
    - design
    - eslint
    - generators
    - hooks
    - stylelint


  If your pull request introduces a breaking change please append `BREAKING CHANGE:` following type / scope.

  Further Reading:
    - https://www.conventionalcommits.org
    - https://github.com/commitizen/conventional-commit-types/blob/master/index.json
-->

## Motivations

- This PR addresses issue #382 
- As well as [addressing the comment](https://github.com/GetJobber/atlantis/pull/395#issuecomment-717333098) to create a private component `<ButtonDismiss>` after member considerations

## Changes

<!-- https://keepachangelog.com/en/1.0.0/ -->

### Added

- A private component `<ButtonDismiss>`

### Changed

- Swapped out `<Icon>` in favor of `<ButtonDismiss>` for the modal component
- Updated snapshot to reflect the new tree structure

### Deprecated

- <!-- soon-to-be removed features -->

### Removed

- <!-- now removed features -->

### Fixed

- <!-- for any bug fixes -->

### Security

- <!-- in case of vulnerabilities -->

## Testing

<!-- How to test your changes. -->
Please inspect the close button is behaving as expected, as well as visually aligns with expectations.

Example screenshots:

Close button at resting state:
![Screen Shot 2020-10-21 at 8 04 22 PM](https://user-images.githubusercontent.com/10535453/96810383-cbeb6f00-13d8-11eb-877c-d2b4fb1f05b2.png)

Close button at hover state:
![Screen Shot 2020-10-21 at 8 04 29 PM](https://user-images.githubusercontent.com/10535453/96810469-d1e15000-13d8-11eb-9ee6-e82dc05e3b2c.png)

---

[In Atlantis we use Github's built in pull request reviews](https://help.github.com/en/articles/about-pull-request-reviews).

![Random photo of Atlantis](https://loremflickr.com/672/400/atlantis)

chris-at-jobber commented:
@eddysims this approach is basically what I was talking about on the other PR - can you give it a review? Behaviour and visuals look good

kingstonfung commented:
🤔
Expanding on that, could the `Button` instead have a `dismiss` status or a `transparentBackground` setting?

eddysims commented:
I think in this case a `ButtonDismiss` is a better option. I think that making it a private 🤫  component that only really needs an `onClick`.

My worry about adding to the `Button` is that we can then have a transparent dismiss button with any icon or a label.

@kingstonfung would you be comfortable taking that on as part of this? If yes, I can show you how to make a private component.

kingstonfung commented:
@eddysims sure, thanks! let's connect

kingstonfung commented:
(sort of) know what to do, will make an attempt to carry this forward!

kingstonfung commented:
@eddysims @chris-at-jobber : Created a private component as per discussion. Rebased, and updated PR comment to reflect the new changes. Please have a look again, thanks!

kingstonfung commented on packages/components/src/ButtonDismiss/ButtonDismiss.tsx:
```diff
@@ -0,0 +1,22 @@
+import { Button } from "../Button";
+
+interface ButtonDismissProps {
+  onClick?(): void;
```
Not sure if I am TypeScripting this portion correctly.... plmk!

What I want to express is the props is expecting a `Function` for the `onClick` prop

eddysims replied:
yep, if you look at https://atlantis.getjobber.com/guides/how-to-create-a-basic-react-component#interface (might have to scroll to the interface section) it shows what is happening here.

You are saying `onClick` accepts a function and will not return anything back.

kingstonfung commented on packages/components/src/ButtonDismiss/ButtonDismiss.tsx:
```diff
@@ -0,0 +1,22 @@
+
+export function ButtonDismiss({ onClick, ariaLabel }: ButtonDismissProps) {
+  return (
+    <div className={styles.closeButtonWrapper}>
```
I'm not a big fan of this wrapper technique if there's another way to avoid it.

But I wasn't able to assign more `classnames` directly to the `<Button>` component... must be an interface setup I've missed somewhere.

eddysims replied:
We specifically **do not** want to assign `className` to `Button`. We want our components to be as opinionated as possible and allowing `className` lets people do some wild things.

eddysims commented on packages/components/src/ButtonDismiss/ButtonDismiss.mdx:
```diff
@@ -0,0 +1,52 @@
+
+<ComponentStatus stage="pre" responsive="no" accessible="no" />
+
+Write a description about what ButtonDismiss component is trying to achieve.
```
please `Write a description about what ButtonDismiss component is trying to achieve.`

kingstonfung replied:
This is now addressed in be4d592. Thanks!

eddysims reviewed (changes requested):
I think just filling out the `mdx` file. Other then that looking good!

eddysims commented:
@chris-at-jobber want to look at this now? I think the modal looks good like this.

chris-at-jobber commented:
Looks great to me 👍

kingstonfung commented:
Thanks guys! 

I'll touch up the MD file with a proper description and then we should be at a good spot

kingstonfung commented:
@eddysims : PTAL again, I've added the comment and thanks for explaining the TS interface & the classnaming restriction on components (agreeing and makes sense!).
